### PR TITLE
implement uniq in go

### LINKED ISSUE #1: Implement uniq in Go
... so that `go run . < in.txt` does the same as `uniq < in.txt`.

### PR DESCRIPTION
**Pull request by justinmueller05082000**
fixes #1

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+)
+
+func main() {
+	content, readErr := ioutil.ReadAll(os.Stdin)
```
Comment (Al2Klimov) [inline main.go]: You're reading all of the data into memory. What happens if you're running this on a VM w/ 1G RAM, but your in.txt is 2G large?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: main.go (hunk 1 above)
Comment (justinmueller05082000) [reply main.go]: which text file is 2G large?

Comment (Al2Klimov) [reply main.go]: `for i in {1..1000}; do for j in {1..1000}; do for k in {1..1000}; do echo 2 done done done > 2G.txt`

That one.

Comment (justinmueller05082000) [reply main.go]: do I have to use read.Bytes again?

Comment (Al2Klimov) [reply main.go]: Exactly.

File: main.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+			fmt.Fprintln(os.Stderr, readErr)
+			os.Exit(1)
+		}
+
+		if len(content) <= 0 {
+			return
+		}
+
+		trimmedContent := bytes.Trim(content, "\n")
+		newContent = append(newContent, trimmedContent)
```
Comment (Al2Klimov) [inline main.go]: You're still collecting all data in memory.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: main.go (hunk 2 above)
Comment (justinmueller05082000) [reply main.go]: you want me to proccess the data one by one? How am I supposed to to that???

Comment (Al2Klimov) [reply main.go]: By remembering the last printed line and comparing with it.

Comment (justinmueller05082000) [reply main.go]: that sounds way more easy than the way i did ._.

File: main.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+		}
+
+		trimmedContent := bytes.Trim(content, "\n")
+
+		if readErr == io.EOF {
+			break
+		}
+
+		if bytes.Compare(checkEquality,trimmedContent) != 0 {
+			fmt.Printf("%#s\n", trimmedContent)
```
Comment (Al2Klimov) [inline main.go]: Why not just `Stdout.Write(content)`?

File: main.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+		if readErr != nil && readErr != io.EOF {
+			fmt.Fprintln(os.Stderr, readErr)
+			os.Exit(1)
+		}
+
+		if len(content) <= 0 {
+			return
+		}
+
+		trimmedContent := bytes.Trim(content, "\n")
```
Comment (Al2Klimov) [inline main.go]: Why not RTrim?

File: main.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+	for _, c := range contentSplit {
+		if bytes.Compare(c, []byte("")) != 0 {
+			data = append(data, c)
+		}
+	}
+
+	for _, a := range data {
+		fmt.Printf("%s\n", a)
+	} */
```
Comment (Al2Klimov) [inline main.go]: ...

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
`echo -n x |uniq` and `echo -n x |go run .` differ.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
They still differ a bit. Add `|hexdump -C` to each command to see the difference.

**Comment by justinmueller05082000** [comment]
Screenshot from 2020-05-25 13-18-10

**Comment by justinmueller05082000** [comment]
See no difference at all

**Comment by Al2Klimov** [comment]
OK, then it's MacOS..

File: main.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+	for {
+		content, readErr := buf.ReadBytes('\n')
+		if readErr != nil && readErr != io.EOF {
+			fmt.Fprintln(os.Stderr, readErr)
+			os.Exit(1)
+		}
+
+		if len(content) != 0 && content[len(content)-1] != 0xa {
+			content = append(content, 0xa)
+		}
```
Comment (Al2Klimov) [inline main.go]: Please use `'\n'`, not 0xa.

File: main.go [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+	var checkEquality []byte
+
+	for {
+		content, readErr := buf.ReadBytes('\n')
+		if readErr != nil && readErr != io.EOF {
+			fmt.Fprintln(os.Stderr, readErr)
+			os.Exit(1)
+		}
+
+		if len(content) != 0 && content[len(content)-1] != 0xa {
```
Comment (Al2Klimov) [inline main.go]: Isn't there any built-in function to check this?

File: main.go [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+			content = append(content, 0xa)
+		}
+
+		if len(content) <= 0 {
+			return
+		}
```
Comment (Al2Klimov) [inline main.go]: What's this for?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: main.go (hunk 7 above)
Comment (justinmueller05082000) [reply main.go]: there probably is but from "scratch" is always better

**Comment by Al2Klimov** [comment]
Wait! Please test **the exact commands I've written before** for difference!

File: main.go (hunk 7 above)
Comment (Al2Klimov) [reply main.go]: It's always... worse.

**Comment by justinmueller05082000** [comment]
Screenshot from 2020-05-25 13-33-56

There is still no difference.

File: main.go (hunk 7 above)
Comment (justinmueller05082000) [reply main.go]: why am I re-developing the linux command uniq then :p

Comment (Al2Klimov) [reply main.go]: To learn something.

Comment (justinmueller05082000) [reply main.go]: Ok but which built-in function do mean?

Comment (Al2Klimov) [reply main.go]: What are you checking for?

Comment (justinmueller05082000) [reply main.go]: the first conditions checks if there already is a newline and the second checks if the last element in the array is a newline or not

Comment (Al2Klimov) [reply main.go]: In short, you're checking whether content is **suffix**ed with `\n`. Any idea which function I mean?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
* [x] `git reset --soft HEAD~$(git rev-list --count $(git merge-base HEAD master)..HEAD^)`
* [x] `git commit --amend --no-edit`
* [x] `git push -f origin feature/implement-uniq-in-go-1`

**Comment by Al2Klimov** [comment]
Write a test protocol.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
Write a test protocol.

**Comment by justinmueller05082000** [comment]
**Test protocol:**

**toUniq.txt:**

```
a
a
b
b
b
a
a
c
c
c
c
c
```

**Go code:**

**Terminal:** `go run . < toUniq.txt `

**Output:**

```
a
b
a
c
```

**Uniq command:**

**Terminal:**  `uniq < toUniq.txt `

**Output:**

```
a
b
a
c
```


**Conclusion:**
Both results are equal.

**Review by Al2Klimov — APPROVED** [review]

**Pull request merged**